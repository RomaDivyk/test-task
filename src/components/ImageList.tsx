import { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import { MediaCart } from './index';

import { StyledImageList } from '../styled/StyledImageList';

import { getDataFetch } from '../redux/ducks/imagesDuck';

export interface Data {
  id: string;
  img: string;
  comments: string[];
}
export interface FetchData {
  data: Data[];
}

export type StateType = {
  fetchData: FetchData;
  isLoading: boolean;
};

const ImageList = () => {
  const data = useSelector((state: StateType) => state.fetchData.data);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getDataFetch());
  }, [dispatch]);

  return (
    <StyledImageList data-testid="imageListId">
      <MediaCart data={data} />
    </StyledImageList>
  );
};

export default ImageList;
