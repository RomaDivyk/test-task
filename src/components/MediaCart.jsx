import React, { useState } from 'react';

import { Container, Grid, CardMedia } from '@mui/material';

import { ModalWindow } from './index';

import { StyledMediaCart } from '../styled/StyledMediaCart';

const MediaCard = ({ data }) => {
  const [open, setOpen] = useState(false);
  const [id, setId] = useState(null);

  const handleOpen = (event) => {
    setId(event.target.id);
    setOpen(true);
  };
  const handleClose = () => setOpen(false);

  const cards = data.map((element) => (
    <Grid item key={element.id} xs={12} sm={6} md={4}>
      <StyledMediaCart>
        <CardMedia
          onClick={handleOpen}
          id={element.id}
          component="img"
          image={element.img}
          alt={`img doesn't download`}
        ></CardMedia>
      </StyledMediaCart>
    </Grid>
  ));

  return (
    <>
      <Container>
        <Grid container spacing={4}>
          {cards}
        </Grid>
      </Container>

      <ModalWindow id={id} data={data} open={open} onClose={handleClose} />
    </>
  );
};

export default MediaCard;
