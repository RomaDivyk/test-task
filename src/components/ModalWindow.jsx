import { useState } from 'react';
import { Modal, Box, CardMedia, TextField, Button, Container, Grid } from '@mui/material';

import { StyledModalWindow, Line, styleModal } from '../styled/StyledModalWindow';

import { sendData } from '../api';

const ModalWindow = ({ id, data, open, onClose }) => {
  const [modalObj, setModalObj] = useState({
    id: null,
    img: '',
    comments: '',
  });
  const [inputData, setInputData] = useState({
    name: '',
    comment: '',
  });

  if (id !== null) {
    const modalResult = data.filter((element) => element.id === id);

    modalObj.id = modalResult[0].id;
    modalObj.img = modalResult[0].img;
    modalObj.comments = modalResult[0].comments.map((comment, index) => <Line key={index}>{comment}</Line>);
  }

  return (
    <StyledModalWindow>
      <Modal open={open} onClose={onClose}>
        <Container sx={styleModal}>
          <Grid container spacing={2}>
            <Grid item xs={10} sm={6.5} md={6} lg={4.5} xl={4} sx={{ maxHeight: '100%' }}>
              <CardMedia image={modalObj.img} alt="empty" component="img" sx={{ maxHeight: '80%', marginBottom: 3 }} />
              <Box>{modalObj.comments}</Box>
            </Grid>

            <Grid item xs={10} sm={6} md={4}>
              <Box
                sx={{
                  display: 'flex',
                  flexDirection: 'column',
                  width: '100%',
                  height: 'auto',
                }}
              >
                <TextField
                  onChange={(e) => setInputData({ ...inputData, name: e.target.value })}
                  id="filled-basic-name"
                  label="Name"
                  variant="filled"
                  margin="normal"
                />
                <TextField
                  onChange={(e) => setInputData({ ...inputData, comment: e.target.value })}
                  id="filled-basic-comment"
                  label="Comment"
                  variant="filled"
                  margin="normal"
                />
                <Button variant="contained" onClick={() => sendData(inputData)}>
                  Leave a comment{' '}
                </Button>
              </Box>
            </Grid>
          </Grid>
        </Container>
      </Modal>
    </StyledModalWindow>
  );
};

export default ModalWindow;
