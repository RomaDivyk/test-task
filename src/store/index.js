import { configureStore } from "@reduxjs/toolkit";
import consoleSlice from "./console_slice";

const store = configureStore({
  reducer: {
    cons: consoleSlice,
  },
});

export default store;
