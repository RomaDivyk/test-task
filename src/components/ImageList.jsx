/* eslint-disable no-unused-vars */
import { useState, useEffect } from "react";

import { fetchData } from "../api";

import MediaCart from "./MediaCart";

const ImageList = () => {
  const [data, setData] = useState([]);
  const [error, setError] = useState("");

  const url = `https://demo1353770.mockable.io/images`;

  useEffect(() => {
    fetchData(url).then((res) => {
      setData(res);
    });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <>
      {<MediaCart data={data} />}
      {error && <h1>{error}</h1>}
    </>
  );
};

export default ImageList;
