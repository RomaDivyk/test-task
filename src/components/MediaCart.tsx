import { useState } from 'react';

import { Container, Grid, CardMedia } from '@mui/material';

import { ModalWindow } from './index';
import { FetchData } from './ImageList';

import { StyledMediaCart } from '../styled/StyledMediaCart';

const MediaCard = ({ data }: FetchData) => {
  const [open, setOpen] = useState(false);
  const [id, setId] = useState<string>('');

  const handleOpen: React.MouseEventHandler<HTMLImageElement> = (event) => {
    setId((event.target as Element).id);
    setOpen(true);
  };
  const handleClose = (): void => setOpen(false);

  const cards = data.map((element) => (
    <Grid item key={element.id} xs={12} sm={6} md={4}>
      <StyledMediaCart>
        <CardMedia
          onClick={handleOpen}
          id={element.id}
          component="img"
          image={element.img}
          alt={`img doesn't download`}
        ></CardMedia>
      </StyledMediaCart>
    </Grid>
  ));

  return (
    <>
      <Container data-testid="mediaCartId">
        <Grid container spacing={4}>
          {cards}
        </Grid>
      </Container>

      <ModalWindow id={id} data={data} open={open} onClose={handleClose} />
    </>
  );
};

export default MediaCard;
