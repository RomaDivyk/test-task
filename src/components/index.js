/// check

import ImageList from "./ImageList";
import MediaCart from "./MediaCart";
import ModalWindow from "./ModalWindow";

export { ImageList, MediaCart, ModalWindow };
