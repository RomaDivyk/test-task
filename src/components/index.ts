import ImageList from './ImageList';
import MediaCart from './MediaCart';
import ModalWindow from './ModalWindow';
import Form from './auth/Form';
import ReactGridTable from './ReactGridTable';
import Charts from './Charts';

export { ImageList, MediaCart, ModalWindow, Form, ReactGridTable, Charts };
