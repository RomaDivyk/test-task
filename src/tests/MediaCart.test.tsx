import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';

import store from '../redux';

import { MediaCart } from '../components';

describe('testing App component', () => {
  it('renders correctly MediaCart component', () => {
    const wrapper = renderer.create(
      <Provider store={store}>
        <BrowserRouter>
          <MediaCart data={[]} />
        </BrowserRouter>
      </Provider>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
