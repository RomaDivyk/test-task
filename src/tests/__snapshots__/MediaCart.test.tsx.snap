// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing App component renders correctly MediaCart component 1`] = `
Array [
  <div
    className="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 css-480o17-MuiGrid-root"
    />
  </div>,
  <div
    className="sc-dmyCSP"
  />,
]
`;
